File: julia/test/test_csv.jl
#!/usr/bin/env julia

# Test MPC designation conversions against CSV test data

push!(LOAD_PATH, joinpath(@__DIR__, "..", "src"))

using MPCDesignation

function main()
    if length(ARGS) < 1
        println(stderr, "Usage: julia test_csv.jl <prov_unpack_to_pack.csv>")
        exit(1)
    end

    csv_file = ARGS[1]
    total_tests = 0
    passed_tests = 0
    failed_tests = 0

    println("=== MPC Designation Conversion Tests (Julia) ===")
    println()

    open(csv_file) do f
        is_first_line = true
        for line in eachline(f)
            # Skip empty lines
            isempty(line) && continue

            # Skip header row
            if is_first_line
                is_first_line = false
                if startswith(line, "unpacked") || occursin("designation", line)
                    continue
                end
            end

            # Parse CSV line
            parts = split(line, ',', limit=2)
            length(parts) != 2 && continue

            unpacked = parts[1]
            expected_packed = parts[2]

            total_tests += 1

            try
                actual_packed = convert_simple(unpacked)
                if actual_packed == expected_packed
                    passed_tests += 1
                else
                    failed_tests += 1
                    if failed_tests <= 10
                        println("FAIL: pack('$unpacked')")
                        println("      Expected: '$expected_packed'")
                        println("      Got:      '$actual_packed'")
                    end
                end
            catch e
                failed_tests += 1
                if failed_tests <= 10
                    msg = e isa MPCDesignationError ? e.msg : string(e)
                    println("FAIL: pack('$unpacked')")
                    println("      Expected: '$expected_packed'")
                    println("      Got error: $msg")
                end
            end
        end
    end

    println()
    println("=== Conversion Test Results ===")
    println("Total:  $total_tests")
    println("Passed: $passed_tests")
    println("Failed: $failed_tests")

    if failed_tests > 10
        println("(Showing first 10 failures only)")
    end

    if failed_tests > 0
        exit(1)
    end
end

main()

File: fortran/test/test_csv.f90
!> Test MPC designation conversions against CSV test data
program test_csv
    use mpc_designation
    implicit none

    character(len=256) :: csv_file, line
    character(len=80) :: unpacked, expected_packed, actual_packed
    integer :: unit_num, ios, total, passed, failed
    integer :: comma_pos
    real :: start_time, end_time

    if (command_argument_count() < 1) then
        print '(A)', 'Usage: test_csv <prov_unpack_to_pack.csv>'
        stop 1
    end if

    call get_command_argument(1, csv_file)

    total = 0
    passed = 0
    failed = 0

    print '(A)', '=== MPC Designation Conversion Tests (Fortran) ==='
    print '(A)', ''

    call cpu_time(start_time)

    open(newunit=unit_num, file=trim(csv_file), status='old', action='read', iostat=ios)
    if (ios /= 0) then
        print '(A)', 'Cannot open file: ' // trim(csv_file)
        stop 1
    end if

    ! Skip header
    read(unit_num, '(A)', iostat=ios) line

    do
        read(unit_num, '(A)', iostat=ios) line
        if (ios /= 0) exit

        if (len_trim(line) == 0) cycle

        ! Parse CSV line
        comma_pos = index(line, ',')
        if (comma_pos == 0) cycle

        unpacked = line(1:comma_pos-1)
        expected_packed = trim(line(comma_pos+1:))

        total = total + 1

        call clear_error()
        actual_packed = convert_simple(unpacked)

        if (MPCDesignationError) then
            failed = failed + 1
            if (failed <= 10) then
                print '(A)', 'FAIL: pack(' // "'" // trim(unpacked) // "'" // ')'
                print '(A)', '      Expected: ' // "'" // trim(expected_packed) // "'"
                print '(A)', '      Got error: ' // trim(error_message)
            end if
        else if (trim(actual_packed) /= trim(expected_packed)) then
            failed = failed + 1
            if (failed <= 10) then
                print '(A)', 'FAIL: pack(' // "'" // trim(unpacked) // "'" // ')'
                print '(A)', '      Expected: ' // "'" // trim(expected_packed) // "'"
                print '(A)', '      Got:      ' // "'" // trim(actual_packed) // "'"
            end if
        else
            passed = passed + 1
        end if

        ! Progress indicator
        if (mod(total, 100000) == 0) then
            print '(A,I0,A)', 'Processed ', total, ' entries...'
        end if
    end do

    close(unit_num)

    call cpu_time(end_time)

    print '(A)', ''
    print '(A)', '=== Conversion Test Results ==='
    print '(A,I0)', 'Total:  ', total
    print '(A,I0)', 'Passed: ', passed
    print '(A,I0)', 'Failed: ', failed
    print '(A,F8.2,A)', 'Time:   ', (end_time - start_time), ' seconds'

    if (failed > 10) then
        print '(A)', '(Showing first 10 failures only)'
    end if

    if (failed > 0) stop 1

end program test_csv
